## machine-learning-zoomcamp-2024 
# Homework Week 1 - E Ekevall

### Setting up the environment

In [1]:
import pandas as pd
import numpy as np

### Read the csv file, assign to the dataframe laptops and view/explore the dataset

In [2]:
df_laptops = pd.read_csv('https://raw.githubusercontent.com/alexeygrigorev/datasets/master/laptops.csv')

In [3]:
df_laptops

Laptop       Status   Brand  \
0     ASUS ExpertBook B1 B1502CBA-EJ0436X Intel Core...          New    Asus   
1     Alurin Go Start Intel Celeron N4020/8GB/256GB ...          New  Alurin   
2     ASUS ExpertBook B1 B1502CBA-EJ0424X Intel Core...          New    Asus   
3     MSI Katana GF66 12UC-082XES Intel Core i7-1270...          New     MSI   
4     HP 15S-FQ5085NS Intel Core i5-1235U/16GB/512GB...          New      HP   
...                                                 ...          ...     ...   
2155  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2156  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2157  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2158  Razer Book 13 Intel Evo Core i7-1165G7/16GB/1T...  Refurbished   Razer   
2159  Razer Book FHD+ Intel Evo Core i7-1165G7/16GB/...  Refurbished   Razer   

           Model                CPU  RAM  Storage Storage type       GPU  \
0     ExpertBook      Intel Core i5    8      512          SSD       NaN   
1             Go      Intel Celeron    8      256          SSD       NaN   
2     ExpertBook      Intel Core i3    8      256          SSD       NaN   
3         Katana      Intel Core i7   16     1000          SSD  RTX 3050   
4            15S      Intel Core i5   16      512          SSD       NaN   
...          ...                ...  ...      ...          ...       ...   
2155       Blade      Intel Core i7   16     1000          SSD  RTX 3060   
2156       Blade      Intel Core i7   16     1000          SSD  RTX 3070   
2157       Blade      Intel Core i7   32     1000          SSD  RTX 3080   
2158        Book  Intel Evo Core i7   16     1000          SSD       NaN   
2159        Book  Intel Evo Core i7   16      256          SSD       NaN   

      Screen Touch  Final Price  
0       15.6    No      1009.00  
1       15.6    No       299.00  
2       15.6    No       789.00  
3       15.6    No      1199.00  
4       15.6    No       669.01  
...      ...   ...          ...  
2155    17.3    No      2699.99  
2156    17.3    No      2899.99  
2157    17.3    No      3399.99  
2158    13.4   Yes      1899.99  
2159    13.4   Yes      1699.99  

[2160 rows x 12 columns]

In [4]:
df_laptops.head(10)

Laptop Status   Brand  \
0  ASUS ExpertBook B1 B1502CBA-EJ0436X Intel Core...    New    Asus   
1  Alurin Go Start Intel Celeron N4020/8GB/256GB ...    New  Alurin   
2  ASUS ExpertBook B1 B1502CBA-EJ0424X Intel Core...    New    Asus   
3  MSI Katana GF66 12UC-082XES Intel Core i7-1270...    New     MSI   
4  HP 15S-FQ5085NS Intel Core i5-1235U/16GB/512GB...    New      HP   
5  MSI Crosshair 17 C12VF-264XES Intel Core i7-12...    New     MSI   
6  Lenovo Thinkpad E14 Gen 4 AMD Ryzen 5 5625U/8G...    New  Lenovo   
7  ASUS VivoBook 15 F515JA-EJ2882W Intel Core i7-...    New    Asus   
8  Medion Akoya E15415 Intel Core i5-10210U/8GB/2...    New  Medion   
9  HP Victus 16-d1038ns Intel Core i7-12700H/16GB...    New      HP   

        Model            CPU  RAM  Storage Storage type       GPU  Screen  \
0  ExpertBook  Intel Core i5    8      512          SSD       NaN    15.6   
1          Go  Intel Celeron    8      256          SSD       NaN    15.6   
2  ExpertBook  Intel Core i3    8      256          SSD       NaN    15.6   
3      Katana  Intel Core i7   16     1000          SSD  RTX 3050    15.6   
4         15S  Intel Core i5   16      512          SSD       NaN    15.6   
5   Crosshair  Intel Core i7   32     1000          SSD  RTX 4060    17.3   
6    ThinkPad    AMD Ryzen 5    8      256          SSD       NaN    14.0   
7    VivoBook  Intel Core i7    8      512          SSD       NaN    15.6   
8       Akoya  Intel Core i5    8      256          SSD       NaN    15.6   
9      Victus  Intel Core i7   16      512          SSD  RTX 3050    16.1   

  Touch  Final Price  
0    No      1009.00  
1    No       299.00  
2    No       789.00  
3    No      1199.00  
4    No       669.01  
5    No      1699.00  
6    No       909.00  
7    No       809.01  
8    No       519.00  
9    No      1149.00

In [5]:
df_laptops.tail(10)

Laptop       Status  Brand  \
2150  Razer Blade 15 Advanced Model QHD Intel Core i...  Refurbished  Razer   
2151  Razer Blade 15 Advanced Model QHD Intel Core i...  Refurbished  Razer   
2152  Razer Blade 15 Base Model FHD Intel Core i7-10...  Refurbished  Razer   
2153  Razer Blade 15 Base Model FHD Intel Core i7-10...  Refurbished  Razer   
2154  Razer Blade 17 4K 144Hz Intel Core i9-12900H/3...  Refurbished  Razer   
2155  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished  Razer   
2156  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished  Razer   
2157  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished  Razer   
2158  Razer Book 13 Intel Evo Core i7-1165G7/16GB/1T...  Refurbished  Razer   
2159  Razer Book FHD+ Intel Evo Core i7-1165G7/16GB/...  Refurbished  Razer   

      Model                CPU  RAM  Storage Storage type       GPU  Screen  \
2150  Blade      Intel Core i7   32     1000          SSD  RTX 3080    15.6   
2151  Blade      Intel Core i7   32     1000          SSD  RTX 3080    15.6   
2152  Blade      Intel Core i7   16      512          SSD  RTX 3060    15.6   
2153  Blade      Intel Core i7   16      512          SSD  RTX 3070    15.6   
2154  Blade      Intel Core i9   32     1000          SSD  RTX 3080    17.3   
2155  Blade      Intel Core i7   16     1000          SSD  RTX 3060    17.3   
2156  Blade      Intel Core i7   16     1000          SSD  RTX 3070    17.3   
2157  Blade      Intel Core i7   32     1000          SSD  RTX 3080    17.3   
2158   Book  Intel Evo Core i7   16     1000          SSD       NaN    13.4   
2159   Book  Intel Evo Core i7   16      256          SSD       NaN    13.4   

     Touch  Final Price  
2150    No      3299.99  
2151    No      3399.99  
2152    No      1232.74  
2153    No      1583.39  
2154    No      4699.99  
2155    No      2699.99  
2156    No      2899.99  
2157    No      3399.99  
2158   Yes      1899.99  
2159   Yes      1699.99

In [6]:
df_laptops.columns

Index(['Laptop', 'Status', 'Brand', 'Model', 'CPU', 'RAM', 'Storage',
       'Storage type', 'GPU', 'Screen', 'Touch', 'Final Price'],
      dtype='object')

In [7]:
df_laptops.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2160 entries, 0 to 2159
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Laptop        2160 non-null   object 
 1   Status        2160 non-null   object 
 2   Brand         2160 non-null   object 
 3   Model         2160 non-null   object 
 4   CPU           2160 non-null   object 
 5   RAM           2160 non-null   int64  
 6   Storage       2160 non-null   int64  
 7   Storage type  2118 non-null   object 
 8   GPU           789 non-null    object 
 9   Screen        2156 non-null   float64
 10  Touch         2160 non-null   object 
 11  Final Price   2160 non-null   float64
dtypes: float64(2), int64(2), object(8)
memory usage: 202.6+ KB


In [8]:
df_laptops.count()

Laptop          2160
Status          2160
Brand           2160
Model           2160
CPU             2160
RAM             2160
Storage         2160
Storage type    2118
GPU              789
Screen          2156
Touch           2160
Final Price     2160
dtype: int64

In [9]:
df_laptops.describe()

RAM      Storage       Screen  Final Price
count  2160.000000  2160.000000  2156.000000  2160.000000
mean     15.413889   596.294444    15.168112  1312.638509
std       9.867815   361.220506     1.203329   911.475417
min       4.000000     0.000000    10.100000   201.050000
25%       8.000000   256.000000    14.000000   661.082500
50%      16.000000   512.000000    15.600000  1031.945000
75%      16.000000  1000.000000    15.600000  1708.970000
max     128.000000  4000.000000    18.000000  7150.470000

In [10]:
df_laptops.nunique()

Laptop          2160
Status             2
Brand             27
Model            121
CPU               28
RAM                9
Storage           12
Storage type       2
GPU               44
Screen            29
Touch              2
Final Price     1440
dtype: int64

The dataset has 12 columns and 2160 records.\
3 Columns (Storage type, GPU and Screen) have missing values.

## Answers to Homework Questions

### Q1. Pandas version
What's the version of Pandas that you installed?

In [11]:
pd.__version__

'2.2.2'

### Q2. Records count

How many records are in the dataset?

In [12]:
len(df_laptops.index)

2160

In [13]:
df_laptops.shape[0]

2160

### Q3. Laptop brands
How many laptop brands are presented in the dataset?

In [14]:
df_laptops['Brand'].nunique()

27

### Q4. Missing values

In [15]:
df_laptops.count()

Laptop          2160
Status          2160
Brand           2160
Model           2160
CPU             2160
RAM             2160
Storage         2160
Storage type    2118
GPU              789
Screen          2156
Touch           2160
Final Price     2160
dtype: int64

In [16]:
df_laptops.isnull().sum()

Laptop             0
Status             0
Brand              0
Model              0
CPU                0
RAM                0
Storage            0
Storage type      42
GPU             1371
Screen             4
Touch              0
Final Price        0
dtype: int64

How many columns in the dataset have missing values?

In [17]:
print(df_laptops.isna().any().sum())

3


In [18]:
print((df_laptops.isnull().sum() > 0).sum())

3


### Q5. Maximum final price

What's the maximum final price of Dell notebooks in the dataset?

In [19]:
df_Dell = df_laptops.groupby('Brand').get_group('Dell')
print(int(df_Dell['Final Price'].max()))

3936


## Q6. Median value of Screen

1. Find the median value of Screen column in the dataset.

In [20]:
df_laptops['Screen'].describe()

count    2156.000000
mean       15.168112
std         1.203329
min        10.100000
25%        14.000000
50%        15.600000
75%        15.600000
max        18.000000
Name: Screen, dtype: float64

In [21]:
print(df_laptops['Screen'].median())

15.6


2. Next, calculate the most frequent value of the same Screen column.

In [22]:
df_laptops['Screen'].mode()

0    15.6
Name: Screen, dtype: float64

3. Use fillna method to fill the missing values in Screen column with the most frequent value from the previous step.

In [23]:
df_screen = df_laptops['Screen'].fillna(15.6)

In [24]:
df_screen.describe()

count    2160.000000
mean       15.168912
std         1.202357
min        10.100000
25%        14.000000
50%        15.600000
75%        15.600000
max        18.000000
Name: Screen, dtype: float64

4. Now, calculate the median value of Screen once again.\
Has it changed?

In [25]:
print(df_screen.median())

15.6


No

# Q7. Sum of weights

1. Select all the "Innjoo" laptops from the dataset.

In [26]:
df_Innjoo = df_laptops.groupby('Brand').get_group('Innjoo')
df_Innjoo

Laptop Status   Brand Model  \
1478  InnJoo Voom Excellence Intel Celeron N4020/8GB...    New  Innjoo  Voom   
1479  InnJoo Voom Excellence Pro Intel Celeron N4020...    New  Innjoo  Voom   
1480  Innjoo Voom Intel Celeron N3350/4GB/64GB eMMC/...    New  Innjoo  Voom   
1481  Innjoo Voom Laptop Max Intel Celeron N3350/6GB...    New  Innjoo  Voom   
1482  Innjoo Voom Laptop Pro Intel Celeron N3350/6GB...    New  Innjoo  Voom   
1483  Innjoo Voom Pro Intel Celeron N3350/6GB/128GB ...    New  Innjoo  Voom   

                CPU  RAM  Storage Storage type  GPU  Screen Touch  Final Price  
1478  Intel Celeron    8      256          SSD  NaN    15.6    No       311.37  
1479  Intel Celeron    8      512          SSD  NaN    15.6    No       392.55  
1480  Intel Celeron    4       64         eMMC  NaN    14.1    No       251.40  
1481  Intel Celeron    6       64         eMMC  NaN    14.1    No       383.61  
1482  Intel Celeron    6      128          SSD  NaN    14.1    No       317.02  
1483  Intel Celeron    6      128         eMMC  NaN    14.1    No       431.38

In [27]:
df_Innjoo.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6 entries, 1478 to 1483
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Laptop        6 non-null      object 
 1   Status        6 non-null      object 
 2   Brand         6 non-null      object 
 3   Model         6 non-null      object 
 4   CPU           6 non-null      object 
 5   RAM           6 non-null      int64  
 6   Storage       6 non-null      int64  
 7   Storage type  6 non-null      object 
 8   GPU           0 non-null      object 
 9   Screen        6 non-null      float64
 10  Touch         6 non-null      object 
 11  Final Price   6 non-null      float64
dtypes: float64(2), int64(2), object(8)
memory usage: 624.0+ bytes


2. Select only columns RAM, Storage, Screen.

In [28]:
X = df_Innjoo.iloc[:, [5, 6, 9]]

3. Get the underlying NumPy array. Let's call it X.

In [29]:
X

RAM  Storage  Screen
1478    8      256    15.6
1479    8      512    15.6
1480    4       64    14.1
1481    6       64    14.1
1482    6      128    14.1
1483    6      128    14.1

In [30]:
X = np.array(X)
X

array([[  8. , 256. ,  15.6],
       [  8. , 512. ,  15.6],
       [  4. ,  64. ,  14.1],
       [  6. ,  64. ,  14.1],
       [  6. , 128. ,  14.1],
       [  6. , 128. ,  14.1]])

4. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.

In [31]:
X_trans = X.T
X_trans

array([[  8. ,   8. ,   4. ,   6. ,   6. ,   6. ],
       [256. , 512. ,  64. ,  64. , 128. , 128. ],
       [ 15.6,  15.6,  14.1,  14.1,  14.1,  14.1]])

In [32]:
XTX = X_trans.dot(X)
XTX

array([[2.52000e+02, 8.32000e+03, 5.59800e+02],
       [8.32000e+03, 3.68640e+05, 1.73952e+04],
       [5.59800e+02, 1.73952e+04, 1.28196e+03]])

In [33]:
XTX = np.matmul(X_trans, X)
XTX

array([[2.52000e+02, 8.32000e+03, 5.59800e+02],
       [8.32000e+03, 3.68640e+05, 1.73952e+04],
       [5.59800e+02, 1.73952e+04, 1.28196e+03]])

5. Compute the inverse of XTX.

In [34]:
XTX_inv = (np.linalg.inv(XTX))
XTX_inv

array([[ 2.78025381e-01, -1.51791334e-03, -1.00809855e-01],
       [-1.51791334e-03,  1.58286725e-05,  4.48052175e-04],
       [-1.00809855e-01,  4.48052175e-04,  3.87214888e-02]])

6. Create an array y with values [1100, 1300, 800, 900, 1000, 1100].

In [35]:
y = np.array([1100, 1300, 800, 900, 1000, 1100])
y

array([1100, 1300,  800,  900, 1000, 1100])

7. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.

In [36]:
z = np.matmul(XTX_inv, X_trans)
w = np.matmul(z, y)
w

array([45.58076606,  0.42783519, 45.29127938])

In [37]:
z = XTX_inv.dot(X_trans)
w = z.dot(y)
w

array([45.58076606,  0.42783519, 45.29127938])

8. What's the sum of all the elements of the result?

In [38]:
answer = sum(w)
print(f"{answer:0.2f}")

91.30
